### ISSUE TITLE
Extend the LinqService

### ISSUE THREAD

**Issue by Franki1986** [linq2db/linq2db#1949]
Hi,

I would like to extend the methods provided of the LinqService.
Is this somehow possible?

```
using(var db = new MyServiceContext(binding, endpointAddress )){
   db.CallMyMethod(); 
}
```

If I understand I have to get the underlying channel, but this one does not include my Interface.
I also don't want to recreate the given LinqService class.

**Comment by sdanyliv** [comment]
As I know, you do not need to extend `linq2db` to do that. You have to create additional service for WCF and configure host. Maybe @MaceWindu can help here, I have used WCF many years ago and forgot everything.

**Comment by MaceWindu** [comment]
I see two options here:
- just implement it as a separate service with own client and server
- introduce additional interface with new methods and implement it in server and client classes, inherited from default server and client classes `LinqService` and `ServiceModelDataContext` (`MyServiceContext` in your case)

**Comment by MaceWindu** [comment]
BTW, unrelated question: have you tried 3.0.0 preview already? Looks like you are one of few remote context users, so it is important to have your feedback before final release, as we already made a lot of changes to it in 3.0.

**Comment by sdanyliv** [comment]
3.0 preview already in Nuget. Why not use it. It is fully functional version with a lot of improvements.
https://www.nuget.org/packages/linq2db/3.0.0-preview.1

**Comment by Franki1986** [comment]
Is [#1908](https://github.com/linq2db/linq2db/issues/1908) implemented in version 3.0 ? This is what I was trying to do with the current opened issue?

**Comment by MaceWindu** [comment]
Nope. I want to look at remote context related tasks after preview 2 released (planned for next week).

For now I have following tasks in mind:
- make remote context first-class citizen for .net core builds
- review queries, that currently fail in 3.0. There are several types of queries, that produce "select *" sql, which is not correct  anyway, as it leads to ineffective queries. With 3.0 changes they currently fail with remote context, because now we need to know all selected columns to transfer data from server to client.
- look at batching issue you linked

Actually I thought about batching yesterday - how it should look. I think it should be something Task-based, where in batch scope we allow only calls with async signature and after we commit batch to server and get results back - we resolve tasks. Need to think about it more, but I think it's correct direction for this feature.

**Comment by jogibear9988** [comment]
I‘ve one question, with remote context, what is serialized? The expression or the SQL?
As atm, we use Interlinq for remote queries...

**Comment by MaceWindu** [comment]
https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/ServiceModel/LinqServiceSerializer.cs

- from client to server: query [AST](https://github.com/linq2db/linq2db/tree/master/Source/LinqToDB/SqlQuery)  https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/ServiceModel/LinqServiceSerializer.cs#L568

- from server to client: resulting data https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/ServiceModel/LinqServiceSerializer.cs#L2044 https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/ServiceModel/LinqServiceResult.cs

**Comment by MaceWindu** [comment]
For 3.0 results serialization changed to perform provider-specific conversions on server-side before sending it to client. Currently it is missing, which creates issues for providers that use such conversions

**Comment by Franki1986** [comment]
What about making a public static serialization access? This would help a lot! People would not have to use other libraries for this like [Interlinq](https://github.com/dotnetprojects/Interlinq-2) (even if I appreciate your work @jogibear9988 ) [Serialize.Linq](https://github.com/esskar/Serialize.Linq) or [Remote.Linq](https://github.com/6bee/Remote.Linq).
We have all we need in LinqToDb for creating serializable remote access.

**Comment by sdanyliv** [comment]
This serializer serializes SQL AST, not Expression Tree. Do you need that?

**Comment by Franki1986** [comment]
Yes I know. In my situation it would help with creating multiple individual sql statements without reconnecting to the server (count, list, sum and so on).
If it is a simpler way to write something like [ExecuteReader](https://github.com/linq2db/linq2db/blob/5e26fa1693208b92b0791dff445ab7b42e79fb1b/Source/LinqToDB/ServiceModel/LinqService.cs#L126) as a ExecuteMultiple, then this would also work.

**Comment by sdanyliv** [comment]
One question. Does Firebird supports returning multiple recordsets per single request? Not related to this thread by i have to keep in mind.

**Comment by Franki1986** [comment]
Unfortunately no..

**Comment by Franki1986** [comment]
One general question, what is the advantage of sql ast serialization infront of Expression tree serialization?

**Comment by sdanyliv** [comment]
One that i can imagine that if you have `SomeTableClass`, you have to share it between server and client. All extension methods also has to be present on server side.
As i know @it, tried Expression Tree serialization first but then he chosen AST serialization, who knows how many problems he has found in this path.

**Comment by Franki1986** [comment]
@jogibear9988 : do you have some examples how you use it with linq2db in the way it works like with the RemoteDataContext? Or do you only use it with simple expressions. I have used Remote.Linq it also worked with anonymous types.

**Comment by jogibear9988** [comment]
I can also use anonymous Types, joins,... interlinq allows serializeing of them.
I use Json.Net for Serialization over a websocket connection.

**Comment by jogibear9988** [comment]
I've now create a simple interlinq sample

https://github.com/dotnetprojects/Interlinq-2/commit/7509d7fe38c33f29728caef226186b4a3fdc109f

**Comment by jogibear9988** [comment]
to query with linq2db you need to modify this class in the sample app: https://github.com/dotnetprojects/Interlinq-2/blob/master/Samples/Sample.Server/SampleQueryHandler.cs

**Comment by jogibear9988** [comment]
@Franki1986 did you have a look?

**Comment by Franki1986** [comment]
Thanks. Yes.. Seams to work fine.. Now I am trying to extend it to call the extension like LoadWith or join.
I also am parallaly trying Remote. Linq. 
Both have similarities.

**Comment by jogibear9988** [comment]
When I see the Code, I much more like the Remote.Linq Syntax. Maybe we should switch to that too.

Maybe we should create a Extension for Remote.Linq to allow Linq2Db special types like LoadWith, Update, ...

**Comment by sdanyliv** [comment]
Guys, why do not create new RemoteContext? Why you are playing with these libraries?

**Comment by jogibear9988** [comment]
We do not want to need the Mapping and Linq2Db on Client side. And also it is a grown Application, where we would not do this huge change...

**Comment by sdanyliv** [comment]
So instead of using linq2db which has no dependencies, you use other library :)

**Comment by jogibear9988** [comment]
No, we also have Our Mappings in 5 or more different assemblies. 
And it is grown Code...

**Comment by jogibear9988** [comment]
And we also don't only access Database Objects, we use the same for querying backend in Memory collections

**Comment by Franki1986** [comment]
Yes that were my thoughts to.. I also like the dynamic serialization with the aqua core. Basically the ef extension class is a good example

**Comment by Franki1986** [comment]
> Guys, why do not create new RemoteContext? Why you are playing with these libraries?

For me it is more a playground to See what are the differences, looking for Performance and usability. But like I said when I opened the issue, it would be nice if I could easily extend the Linq Service ;-)

**Comment by jogibear9988** [comment]
I think we will remove it in the long turn. As we switched to a "HTML" ui in the last 2 years. I first thought there to build the expression tree in javascript (like interlinq serializes to json), but we now use dynamic.linq, and some string querys in the frontend. Joins are not possible with this solution at the moment, but for now we don't need them

**Comment by sdanyliv** [comment]
Oh man, interesting things here. If we just copy Interlinq code to our repository and rewrite remoting. So many issues may gone if library works as expected.
And, for example, divide library to abstraction `linq2db.Core` (our linq extensions) `linq2db.Remote` and concrete implementation - `linq2db`, and transport layer `linq2db.HTTP.Client`, `linq2db.HTTP.Server`

Just thinking how it can be done. No free time for realization (
Maybe 4.0 if you help with this.

**Comment by jogibear9988** [comment]
i will look bit into it, also to see wich is the cleanest library as a base.

**Comment by jogibear9988** [comment]
and the fastest, and wich one supports most, ...

**Comment by Franki1986** [comment]
I have played a little bit with Remote.Linq and it works pretty well for me.
I started making an extension for my self:

[Remote.Linq.LinqToDB](https://github.com/Franki1986/Remote.Linq.LinqToDB)

**Comment by sdanyliv** [comment]
Maybe i’m wrong but looks like Interlinq is smarter in serializing custom methods.

**Comment by Franki1986** [comment]
in which kind, part of coding implementation or Performance?

**Comment by sdanyliv** [comment]
Just analyzed your sources. What are you trying to do with `LoadWith`?
I'm interesting how both libraries will transfer for example `Sql.ToNullable()`

**Comment by Franki1986** [comment]
It is just a loadwith extension for a IRemoteQueryable.. On server side I visit the custom expresions. I am not 100% sure if this can be done without..

**Comment by sdanyliv** [comment]
Try something like this:
```cs
var query = 
   from p in db.GetTable<Person>()
   from a in db.GetTable<Article>().LeftJoin(a => a.PersonId == p.PersonId)
   select new
   {
        p.Name,
        ArticleId = Sql.ToNullable(a.Id)
   }
```
I'm interested how LeftJoin and Sql.ToNullable are transferred.

**Comment by Franki1986** [comment]
I will report

**Comment by jogibear9988** [comment]
I think we also need to think about how we can do security, see also: https://github.com/6bee/Remote.Linq/issues/52
But this is the same problem with Interlinq

**Comment by jogibear9988** [comment]
I also think interlinq needs work. It works on our side, but some of my changes I did years ago seem hacky to me...

for example this:
https://github.com/dotnetprojects/Interlinq-2/commit/92ccf0a39689744b6499a9cb36055997fcad3f99
(what about hash collission, and why did I do this??)

This also looks hacky...
https://github.com/dotnetprojects/Interlinq-2/commit/75a58889147c154c52ce95af07d4076969c02849

**Comment by sdanyliv** [comment]
@jogibear9988, security is doable by introducing entities filters and CRUD operation handlers.
We can introduce handler method for example

```cs
public IQueryble<Person> AccessFilter(ITable<Person> table)
{
   if (!CurrentUserIsAdmin)
       return table.Where(bla bla)
   return table
}
```
And if such method is defined in handler - we can dynamically call it before generating SQL.

**Comment by jogibear9988** [comment]
@sdanyliv thats no problem, I' ve done it like this.
The Problem is the one's mentioned in the issue, that you can use Directory class for example....

But that can also be fixed by a expression visitor I think.

**Comment by jogibear9988** [comment]
I've now converted Interlinq to a NetStandard2.1 Project (and Net4 Project), but not yet tested.
Also Silverlight Code (and more) removed.
I think Anonymous Types will still be a Problem, cause this does not work in NetStandard: https://github.com/dotnetprojects/Interlinq-2/blob/master/InterLinq/Types/Anonymous/DynamicAssemblyHolder.cs

**Comment by sdanyliv** [comment]
@jogibear9988, we can always move to https://github.com/castleproject/Core for anonymous classes creation. It is how EF Core works now in creating proxy objects for Change Tracker.

**Comment by jogibear9988** [comment]
Oh, I see, maybe I can change the asseblybuilder like in castle core:

https://github.com/castleproject/Core/blob/e2dfb57020d9dbb4b31f3ce548b34cb35ffa3307/src/Castle.Core/Components.DictionaryAdapter/DictionaryAdapterFactory.cs#L169

**Comment by sdanyliv** [comment]
Who knows :)
I have high load now on my primary job and can not spend a lot of time on experiments. There are more prioritized tasks to do.

**Comment by jogibear9988** [comment]
Fixed... Now the Code for NetStandard and Net45 are the same, so it should work.

**Comment by Franki1986** [comment]
> Try something like this:
> 
> ```cs
> var query = 
>    from p in db.GetTable<Person>()
>    from a in db.GetTable<Article>().LeftJoin(a => a.PersonId == p.PersonId)
>    select new
>    {
>         p.Name,
>         ArticleId = Sql.ToNullable(a.Id)
>    }
> ```
> 
> I'm interested how LeftJoin and Sql.ToNullable are transferred.

This is the textual visualization of the remote linq expression that is transfered:

`System.Linq.Queryable.SelectMany(Remote.Linq.DynamicQuery.QueryableResourceDescriptor, Quote(p => LinqToDB.LinqExtensions.LeftJoin(New Remote.Linq.DynamicQuery.VariableQueryArgument`1[System.Linq.IQueryable`1[IacCentralDataBaseMapping.Data.Remote_Linq.Article]](Remote.Linq.DynamicQuery.QueryableResourceDescriptor).Value, Quote(a => (a.PersonId Equal p.ID)))), Quote((p,a) => New <>f__AnonymousType2`2[System.String,System.Nullable`1[System.Int32]](p.Name, LinqToDB.Sql.ToNullable(a.ID))))`


Client and server produce the same end result.

**Comment by jogibear9988** [comment]
I like it.
For me it looks much easier then Interlinq

Need to try if it is also usable without WCF and custom serializer

**Comment by Franki1986** [comment]
> BTW, unrelated question: have you tried 3.0.0 preview already? Looks like you are one of few remote context users, so it is important to have your feedback before final release, as we already made a lot of changes to it in 3.0.

A short update, until now the remote context works like expected!
